File: src/main/java/com/babylonml/backend/training/operations/SoftMaxCrossEntropyCostFunction.java
package com.babylonml.backend.training.operations;

import com.babylonml.backend.cpu.TensorOperations;
import com.babylonml.backend.training.execution.TensorPointer;
import com.babylonml.backend.training.execution.TrainingExecutionContext;
import com.babylonml.backend.cpu.MatrixOperations;
import com.babylonml.backend.cpu.VectorOperations;
import jdk.incubator.vector.FloatVector;
import jdk.incubator.vector.VectorOperators;
import jdk.incubator.vector.VectorSpecies;
import org.jetbrains.annotations.NotNull;
import org.jspecify.annotations.NonNull;

public final class SoftMaxCrossEntropyCostFunction extends AbstractOperation implements CostFunction {
    private static final VectorSpecies<Float> SPECIES = FloatVector.SPECIES_PREFERRED;

    private final int @NonNull [] maxShape;

    private TensorPointer softMaxResultPointer;
    private TensorPointer expectedProbabilityPointer;

    private final boolean requiresDerivativeChainValue;

    private boolean trainingMode;

    public SoftMaxCrossEntropyCostFunction(@NonNull Operation expectedProbability,
                                           @NonNull Operation predictedOperation) {
        super(predictedOperation, expectedProbability);

        this.maxShape = TensorOperations.calculateMaxShape(predictedOperation.getMaxResultShape(),
                expectedProbability.getMaxResultShape());

        this.requiresDerivativeChainValue = predictedOperation.requiresBackwardDerivativeChainValue();
    }

    @Override
    public int @NonNull [] getMaxResultShape() {
        return maxShape;
    }

    @Override
    public @NonNull TensorPointer forwardPassCalculation() {
        var predictedOperandResultPointer = leftOperation.forwardPassCalculation();
        var predictedOperandBuffer = predictedOperandResultPointer.buffer();
        var predictedOperandOffset = predictedOperandResultPointer.offset();


        softMaxResultPointer = executionContext.allocateForwardMemory(this, predictedOperandResultPointer.shape());
        expectedProbabilityPointer = rightOperation.forwardPassCalculation();

        var softMaxBuffer = softMaxResultPointer.buffer();
        var softMaxOffset = softMaxResultPointer.offset();

        var expectedProbability = expectedProbabilityPointer.buffer();
        var expectedProbabilityOffset = expectedProbabilityPointer.offset();

        var shape = predictedOperandResultPointer.shape();
        if (shape.length != 2) {
            throw new IllegalArgumentException("Softmax cross entropy cost function only supports 2D tensors");
        }

        MatrixOperations.softMaxByRows(predictedOperandBuffer, predictedOperandOffset, shape[0], shape[1],
                softMaxBuffer, softMaxOffset);

        if (trainingMode) {
            return TrainingExecutionContext.NULL;
        }

        var stride = TensorOperations.stride(predictedOperandResultPointer.shape());
        var loopBound = SPECIES.loopBound(stride);
        var vecSum = FloatVector.zero(SPECIES);
        for (int i = 0; i < loopBound; i += SPECIES.length()) {
            var vec = FloatVector.fromArray(SPECIES, softMaxBuffer,
                    softMaxOffset + i).lanewise(VectorOperators.LOG);
            var expectedVec = FloatVector.fromArray(SPECIES, expectedProbability, i + expectedProbabilityOffset);
            vecSum = vec.fma(expectedVec, vecSum);
        }

        var sum = vecSum.reduceLanes(VectorOperators.ADD);
        for (int i = loopBound; i < stride; i++) {
            sum += (float) Math.log(softMaxBuffer[softMaxOffset + i]) * expectedProbability[i + expectedProbabilityOffset];
        }

        var result = executionContext.allocateForwardMemory(this, 1, 1);

        var resultBuffer = result.buffer();
        var resultOffset = result.offset();

        resultBuffer[resultOffset] = -sum;

        return result;
    }

    @Override
    public @NonNull TensorPointer leftBackwardDerivativeChainValue() {
        var softMaxBuffer = softMaxResultPointer.buffer();
        var softMaxOffset = softMaxResultPointer.offset();

        var expectedProbability = expectedProbabilityPointer.buffer();
        var expectedProbabilityOffset = expectedProbabilityPointer.offset();

        var result = executionContext.allocateBackwardMemory(this, softMaxResultPointer.shape());
        var resultBuffer = result.buffer();
        var resultOffset = result.offset();

        var stride = TensorOperations.stride(softMaxResultPointer.shape());
        VectorOperations.subtractVectorFromVector(softMaxBuffer, softMaxOffset, expectedProbability,
                expectedProbabilityOffset, resultBuffer, resultOffset, stride);

        return result;
    }

    @Override
    public @NonNull TensorPointer rightBackwardDerivativeChainValue() {
        return TrainingExecutionContext.NULL;
    }

    @NotNull
    @Override
    public int @NonNull [][] getForwardMemoryAllocations() {
        return new int[][]{
                maxShape,
                new int[]{1, 1}
        };
    }

    @Override
    public int @NonNull [][] getBackwardMemoryAllocations() {
        return new int[][]{
                maxShape
        };
    }

    @Override
    public boolean requiresBackwardDerivativeChainValue() {
        return requiresDerivativeChainValue;
    }

    @Override
    public void trainingMode() {
        trainingMode = true;
    }

    @Override
    public void fullPassCalculationMode() {
        trainingMode = false;
    }
}

File: src/main/kotlin/com/babylonml/backend/training/operations/SoftMaxCrossEntropyCostFunction.kt
package com.babylonml.backend.training.operations

import com.babylonml.backend.common.CommonTensorOperations
import com.babylonml.backend.cpu.MatrixOperations
import com.babylonml.backend.cpu.VectorOperations
import com.babylonml.backend.training.execution.TensorPointer
import com.babylonml.backend.training.execution.TrainingExecutionContext
import it.unimi.dsi.fastutil.ints.IntImmutableList
import jdk.incubator.vector.FloatVector
import jdk.incubator.vector.VectorOperators
import jdk.incubator.vector.VectorSpecies
import kotlin.math.ln

class SoftMaxCrossEntropyCostFunction(
    expectedProbability: Operation,
    predictedOperation: Operation
) : AbstractOperation(predictedOperation, expectedProbability), CostFunction {
    override val maxResultShape: IntImmutableList =
        CommonTensorOperations.calculateMaxShape(
            predictedOperation.maxResultShape,
            expectedProbability.maxResultShape
        )

    private var softMaxResultPointer: TensorPointer? = null
    private var expectedProbabilityPointer: TensorPointer? = null

    private var trainingMode = false

    override fun forwardPassCalculation(): TensorPointer {
        val predictedOperandResultPointer = leftPreviousOperation!!.forwardPassCalculation()
        val predictedOperandBuffer = predictedOperandResultPointer.buffer()
        val predictedOperandOffset = predictedOperandResultPointer.offset()


        softMaxResultPointer = executionContext.allocateForwardMemory(
            this,
            predictedOperandResultPointer.shape
        )
        expectedProbabilityPointer = rightPreviousOperation!!.forwardPassCalculation()

        val softMaxBuffer = softMaxResultPointer!!.buffer()
        val softMaxOffset = softMaxResultPointer!!.offset()

        val expectedProbability = expectedProbabilityPointer!!.buffer()
        val expectedProbabilityOffset = expectedProbabilityPointer!!.offset()

        val shape = predictedOperandResultPointer.shape
        require(shape.size == 2) { "Softmax cross entropy cost function only supports 2D tensors" }

        MatrixOperations.softMaxByRows(
            predictedOperandBuffer, predictedOperandOffset, shape.getInt(0),
            shape.getInt(1),
            softMaxBuffer, softMaxOffset
        )

        if (trainingMode) {
            return TrainingExecutionContext.NULL
        }

        val stride = CommonTensorOperations.stride(
            predictedOperandResultPointer.shape
        )
        val loopBound = SPECIES.loopBound(stride)
        var vecSum = FloatVector.zero(SPECIES)
        run {
            var i = 0
            while (i < loopBound) {
                val vec = FloatVector.fromArray(
                    SPECIES, softMaxBuffer,
                    softMaxOffset + i
                ).lanewise(VectorOperators.LOG)
                val expectedVec = FloatVector.fromArray(SPECIES, expectedProbability, i + expectedProbabilityOffset)
                vecSum = vec.fma(expectedVec, vecSum)
                i += SPECIES.length()
            }
        }

        var sum = vecSum.reduceLanes(VectorOperators.ADD)
        for (i in loopBound until stride) {
            sum += ln(softMaxBuffer[softMaxOffset + i].toDouble()).toFloat() * expectedProbability[i + expectedProbabilityOffset]
        }

        val result = executionContext.allocateForwardMemory(this, IntImmutableList.of(1, 1))

        val resultBuffer = result.buffer()
        val resultOffset = result.offset()

        resultBuffer[resultOffset] = -sum

        return result
    }

    override fun leftBackwardDerivativeChainValue(): TensorPointer {
        val softMaxBuffer = softMaxResultPointer!!.buffer()
        val softMaxOffset = softMaxResultPointer!!.offset()

        val expectedProbability = expectedProbabilityPointer!!.buffer()
        val expectedProbabilityOffset = expectedProbabilityPointer!!.offset()

        val result = executionContext.allocateBackwardMemory(this, softMaxResultPointer!!.shape)
        val resultBuffer = result.buffer()
        val resultOffset = result.offset()

        val stride = CommonTensorOperations.stride(
            softMaxResultPointer!!.shape
        )
        VectorOperations.subtractVectorFromVector(
            softMaxBuffer, softMaxOffset, expectedProbability,
            expectedProbabilityOffset, resultBuffer, resultOffset, stride
        )

        return result
    }

    override fun rightBackwardDerivativeChainValue(): TensorPointer {
        return TrainingExecutionContext.NULL
    }

    override val forwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(
            maxResultShape,
            IntImmutableList.of(1, 1)
        )

    override val backwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(maxResultShape)

    override fun trainingMode() {
        trainingMode = true
    }

    override fun fullPassCalculationMode() {
        trainingMode = false
    }

    override val requiresBackwardDerivativeChainValue: Boolean by lazy {
        leftPreviousOperation!!.requiresBackwardDerivativeChainValue ||
                rightPreviousOperation!!.requiresBackwardDerivativeChainValue
    }

    companion object {
        private val SPECIES: VectorSpecies<Float> = FloatVector.SPECIES_PREFERRED
    }
}
